### PR TITLE
Sort the ranks of nodes by the switches.

### PR DESCRIPTION
**Pull request by workingloong**
### What changes were proposed in this pull request?

Sort the ranks of nodes by the switches.

### Why are the changes needed?

In a DDP job, the communication packets between nodes with continuous ranks under an asw will not pass the psw.

### Does this PR introduce any user-facing change?

No.

### How was this patch tested?

UT.

### CODE REVIEW

**Review by samplise — APPROVED** [review]

File: dlrover/python/master/elastic_training/net_topology.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+# limitations under the License.
+
+from abc import ABCMeta, abstractmethod
+from collections import OrderedDict
+from dataclasses import dataclass
+from typing import Dict, List, Tuple
+
+
+@dataclass
+class NodeTopologyMeta(object):
```
Comment (BalaBalaYi) [inline dlrover/python/master/elastic_training/net_topology.py:21]: The py file's name better named to 'topology'. The topology is not just about the network, but also other env resource. The word 'topology' is more abstract for using.

File: dlrover/python/master/elastic_training/net_topology.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    node_rank: int = 0
+    process_num: int = 0
+    node_ip: str = ""
+    asw: str = ""
+    psw: str = ""
+
+
+class TopologyQuerier(metaclass=ABCMeta):
+    @abstractmethod
+    def query(self, node_ip) -> Tuple[str, str]:
```
Comment (BalaBalaYi) [inline dlrover/python/master/elastic_training/net_topology.py:31]: According to the above, the ```query``` method definitions seems to be too specific in the network topology scenario.  Similar problems for other method definition.

File: dlrover/python/master/elastic_training/net_topology.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+        """Query the asw and psw id of a node by the IP."""
+        pass
+
+
+class DefaultTopologyQuerier(TopologyQuerier):
+    def query(self, node_ip) -> Tuple[str, str]:
+        return "", ""
+
+
+class DpTopologySorter(TopologySorter):
```
Comment (BalaBalaYi) [inline dlrover/python/master/elastic_training/net_topology.py:50]: What does 'Dp' means?

File: dlrover/python/master/elastic_training/net_topology.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    """
+    The sorter places the nodes under an asw (access switch) together in
+    the list of nodes. In allreduce communication, the communication packets
+    between nodes with continuous ranks under an asw will not pass the psw.
+
+    """
+
+    def sort(
```
Comment (BalaBalaYi) [inline dlrover/python/master/elastic_training/net_topology.py:58]: I'm not sure. Is the following implementation adapt most network arch on kubernetes?

File: dlrover/python/master/elastic_training/net_topology.py [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+        self, nodes: Dict[int, NodeTopologyMeta]
+    ) -> Dict[int, NodeTopologyMeta]:
```
Comment (BalaBalaYi) [inline dlrover/python/master/elastic_training/net_topology.py:60]: I don't know why need the following part :
```
asw0_nodes = asw_nodes.pop(rank0_asw, [])
        for node_meta in asw0_nodes:
            sorted_nodes[node_meta.node_rank] = node_meta
```

So i did a merge:

```
asw_nodes: Dict[str, List[NodeTopologyMeta]] = {}

# setup asw nodes
for _, meta in nodes.items():
    asw_nodes.setdefault(meta.asw, [])
    asw_nodes[meta.asw].append(meta)

# init 1st element
rank0_node = nodes.values()[0] 
rank0_asw = rank0_node.asw

# sort nodes 
sorted_nodes: Dict[int, NodeTopologyMeta] = OrderedDict()
for node_metas in asw_nodes.values():
    for node_meta in node_metas:
        sorted_nodes[node_meta.node_rank] = node_meta

return sorted_nodes
```

File: dlrover/python/master/elastic_training/rdzv_manager.py [hunk 6]
```diff
@@ -54,18 +60,23 @@ def __init__(self):
...
         self._node_unit = 1
         self._name = ""
         self._latest_rdzv_nodes = []
         self._start_rdzv_ts = 0
+        # key is the node rank, value is the time.
         self._node_rdzv_times: Dict[int, int] = {}
         self._latest_log_nodes_time = 0
+        # key is the node rank, value is the step.
         self._save_ckpt_nodes: Dict[int, int] = {}
+        self._topology_querier = DefaultTopologyQuerier()
```
Comment (BalaBalaYi) [inline dlrover/python/master/elastic_training/rdzv_manager.py:78]: Better use a 'topology manager' to include the relevant implementation.

File: dlrover/python/master/servicer.py [hunk 7]
```diff
@@ -259,8 +257,8 @@ def _get_comm_world(self, request: grpc.CommWorldRequest):
         res = grpc.RendezvousState(world={})
         res.group = group
         res.round = rdzv_round
-        for rank_id, worker_num in nodes.items():
-            res.world[rank_id] = worker_num
+        for rank_id, meta in nodes.items():
```
Comment (BalaBalaYi) [inline dlrover/python/master/servicer.py:260]: ```meta``` -> ```topology_meta```

File: dlrover/python/master/elastic_training/net_topology.py (hunk 2 above)
Comment (BalaBalaYi) [reply dlrover/python/master/elastic_training/net_topology.py:31]: Also, for current method, should be named ```queryAswAndPsw```.

File: dlrover/python/master/elastic_training/rdzv_manager.py [hunk 8]
```diff
@@ -198,7 +211,15 @@ def join_rendezvous(
                 self._start_rdzv_ts = time.time()
             if node_rank in self._waiting_nodes:
                 return self._rdzv_round
-            self._waiting_nodes[node_rank] = local_world_size
+            asw, psw = self._topology_querier.query(node_ip)
+            meta = NodeTopologyMeta(
```
Comment (BalaBalaYi) [inline dlrover/python/master/elastic_training/rdzv_manager.py:215]: Better to distinguish the naming of local variables 'meta'. Consider there is so many concepts of metadata.

File: dlrover/python/master/elastic_training/rdzv_manager.py [hunk 9]
```diff
@@ -240,7 +261,9 @@ def sync_ckpt_nodes(self, node_id, step):
         return False
 
     @abstractmethod
-    def get_comm_world(self, node_rank):
+    def get_comm_world(
+        self, node_rank
+    ) -> Tuple[int, int, Dict[int, NodeTopologyMeta]]:
```
Comment (BalaBalaYi) [inline dlrover/python/master/elastic_training/rdzv_manager.py:266]: Need annotations for complex tuple strcuture.

**Pull request merged**